### DISCUSSION TITLE
condition: prototype of predicateFn

### DISCUSSION THREAD

**Discussion by Bugslive** [nartc/mapper#492] [Q&A]
Hi all,
I'm trying to use condition but I've got an error "TypeError: transformationPreConditionPredicate is not a function".
On the documentation example, what would be the "source.hasPet" declaration ?

NB: I'm working with classes.

Thanks for your help.

**Comment by nartc** [comment]
Hi @Bugslive, can you provide your mapping configuration?

**Comment by Bugslive** [comment]
I have a classe DTO and a class Model like below
```
class AuthServerDTO {
name: string;
type: "oidc" | "keycloak";
parameters: OidcParameters | KeycloakParameters //classes
}

class AuthServerModel { //model for database
name: string;
type: "oidc" | "keycloak";
parameters: OidcParametersModel | KeycloakParametersModel //classes
}

```
I create a mapper and add the following maps:
```
    createMap(this.mapper, OidcParametersModel , OidcParameters);
    createMap(this.mapper, OidcParameters, OidcParametersModel);
    createMap(this.mapper, KeycloakParametersModel, KeycloakParameters);
    createMap(this.mapper, KeycloakParameters, KeycloakParametersModel);
    createMap(
      this.mapper,
      AuthServer,
      AuthServerModel,
      forMember(
        (AuthServerModel) => AuthServerModel.parameters,
        condition((AuthServer) => {
          return AuthServer.type === 'oidc';
        }),
        mapWith(OidcParametersModel , OidcParameters, (AuthServer) => AuthServer.parameters)
      ),
      forMember(
        (AuthServerModel) => AuthServerModel.parameters,
        condition((AuthServer) => {
          return AuthServer.type === 'keycloak';
        }),
        mapWith(KeycloakParametersModel, KeycloakParameters, (AuthServer) => AuthServer.parameters)
      )
    );
```
For condition function, I've tried:

- ```condition((AuthServer) => {return AuthServer.type === 'oidc';})``` like in the example above
- ```condition((AuthServer) => AuthServer.type === 'oidc')```
- with a method like this in class AuthServerDTO:
    - ```isOidc(): boolean {return this.type === 'oidc';}```

**Reply by nartc** [reply]
That's not how you'd use `MappingConfiguration` and I'd need a better error message for misuses like your case here. Can you tell me what you want to achieve? I can kinda tell you want to map something conditionally but I'd like to hear a more detailed use-case so I can help better.

**Comment by Bugslive** [comment]
My purpose is to ensure that the automapper uses the right mapping for the property parameters.
According to the property 'type', I'm able to "force" it.

**Reply by nartc** [reply]
Can you try this instead?

```ts
// the rest of your code goes here...

createMap(
  this.mapper,
  AuthServer,
  AuthServerModel,
  forMember(
    (AuthServerModel) => AuthServerModel.parameters,
    mapDefer((authServer) => {
        if (authServer.type === 'oidc') 
            return mapWith(OidcParametersModel , OidcParameters, (AuthServer) => AuthServer.parameters);
        return mapWith(KeycloakParametersModel, KeycloakParameters, (AuthServer) => AuthServer.parameters);
    })
  )
);
```

You can also do:

```ts
createMap(
  this.mapper,
  AuthServer,
  AuthServerModel,
  forMember(
    (AuthServerModel) => AuthServerModel.parameters,
    mapDefer((authServer) => {
        let destination = KeycloakParameters;
        let source = KeycloakParametersModel;

        if (authServer.type === 'oidc') {
            destination = OidcParameters;
            source = OidcParametersModel;
        }
        
        return mapWith(source, destination, authServer => authServer.parameters);
    })
  )
);
```

The takeaway is to use `mapDefer` (https://automapperts.netlify.app/docs/mapping-configuration/for-member/map-defer/)

**Comment by Bugslive** [comment]
Thank you very much @nartc  for your help, it works fine
